lib/asset/AssetGenerator.js

```diff
@@ -368,8 +405,55 @@ class AssetGenerator extends Generator {
 	 * @param {Hash} hash hash that will be modified
 	 * @param {UpdateHashContext} updateHashContext context for updating hash
 	 */
-	updateHash(hash, { module }) {
-		hash.update(module.buildInfo.dataUrl ? "data-url" : "resource");
+	updateHash(hash, { module, runtime, runtimeTemplate, chunkGraph }) {
+		if (module.buildInfo.dataUrl) {
+			hash.update("data-url");
+			// this.dataUrlOptions as function should be pure and only depend on input source and filename
+			// therefore it doesn't need to be hashed
+			if (typeof this.dataUrlOptions === "function") {
+				const ident = /** @type {{ ident?: string }} */ (this.dataUrlOptions)
+					.ident;
+				if (ident) hash.update(ident);
+			} else {
+				if (this.dataUrlOptions.encoding)
+					hash.update(this.dataUrlOptions.encoding);
```

Comment:
I think here `Default_encoding` should be used too. e.g. scenario:
1. first run -> no encoding in `dataUrlOptions` -> hash A
2. second run -> set encoding to base64 -> hash B (but it should be the same)

Reply:
ok, but I would then change it to `if (this.dataUrlOptions.encoding && this.dataUrlOptions.encoding !== DEFAULT_ENCODING)` to avoid hashing when possible...